File: users-repo-api/src/main/java/net/plshark/users/repo/RolesRepository.java
package net.plshark.users.repo;

import net.plshark.users.Role;
import reactor.core.publisher.Mono;

/**
 * Repository for saving, deleting, and retrieving roles
 */
public interface RolesRepository {

    /**
     * Get a role by ID
     * @param id the ID
     * @return the matching role
     */
    Mono<Role> getForId(long id);

    /**
     * Get a role by name
     * @param name the role name
     * @return the matching role
     */
    Mono<Role> getForName(String name);

    /**
     * Insert a new role
     * @param role the role to insert
     * @return the inserted role, will have the ID set
     */
    Mono<Role> insert(Role role);

    /**
     * Delete a role by ID
     * @param roleId the role ID
     * @return an empty result
     */
    Mono<Void> delete(long roleId);
}

File: user-error-server/src/main/kotlin/net/plshark/users/repo/RolesRepository.kt
package net.plshark.users.repo

import kotlinx.coroutines.flow.Flow
import net.plshark.users.model.Role
import net.plshark.users.model.RoleCreate

/**
 * Repository for saving, deleting, and retrieving roles
 */
interface RolesRepository {

    /**
     * Find a role by ID
     * @param id the ID
     * @return the matching role or null if not found
     */
    suspend fun findById(id: Long): Role?

    /**
     * Find a role by name
     * @param applicationId the parent application ID
     * @param name the role name
     * @return the matching role or null if not found
     */
    suspend fun findByApplicationIdAndName(applicationId: Long, name: String): Role?

    /**
     * Find all roles up to the maximum result count and starting at an offset
     * @param maxResults the maximum number of results to return
     * @param offset the offset to start the list at, 0 to start at the beginning
     * @return a [Flow] emitting the roles
     */
    fun getRoles(maxResults: Int, offset: Long): Flow<Role>

    /**
     * Find all roles belonging to an application
     * @param applicationId the application ID
     * @return a [Flow] emitting the roles
     */
    fun findRolesByApplicationId(applicationId: Long): Flow<Role>

    /**
     * Insert a new role
     * @param role the role to insert
     * @return the inserted role
     */
    suspend fun insert(role: RoleCreate): Role

    /**
     * Delete a role by ID
     * @param id the role ID
     */
    suspend fun deleteById(id: Long)
}
